File: server/app/services/s3.ts
import { S3Client, PutObjectCommand } from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";
import { S3_FOLDER_STRUCT } from "../types/services.types";

const S3_BUCKET = process.env.S3_BUCKET_NAME;
const s3Client = new S3Client({ region: process.env.AWS_REGION });

const s3 = async (
    folder: S3_FOLDER_STRUCT, 
    contentType: "image/jpeg" | "image/png" | "video/mp4" = "image/jpeg"
): Promise<{ presignedUrl: string; s3Url: string }> => {

    const extension = contentType.split("/")[1];
    const key = `${folder}/${Date.now()}.${extension}`;

    const presignedUrl = await getSignedUrl(s3Client, new PutObjectCommand({
        Bucket: S3_BUCKET,
        Key: key,
        ContentType: contentType,
    }), { expiresIn: 300 });

    return { presignedUrl, s3Url: `https://${S3_BUCKET}.s3.amazonaws.com/${key}` };
};



export default s3;

File: ml/services/s3.py
import os
import time
import boto3
from typing import Tuple, Optional

from dotenv import load_dotenv

load_dotenv()
S3_ML_BUCKET_NAME=os.getenv("S3_ML_BUCKET_NAME")
AWS_REGION = os.getenv("AWS_REGION")

s3_client = boto3.client(
    "s3",
    region_name=AWS_REGION,
    aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID"),
    aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY"),
)

def s3(
    folder: str,
    content_type: str = "image/jpeg",
    filename: Optional[str] = None,
) -> Tuple[str, str]:
    """
        Generate presigned S3 upload URL + public S3 URL
    """

    extension = content_type.split("/")[-1]

    if filename:
        if not filename.endswith(f".{extension}"):
            filename = f"{filename}.{extension}"
        key = f"{folder}/{filename}"
    else:
        key = f"{folder}/{int(time.time() * 1000)}.{extension}"

    presigned_url = s3_client.generate_presigned_url(

        ClientMethod="put_object",

        Params={
            "Bucket": S3_ML_BUCKET_NAME,
            "Key": key,
            "ContentType": content_type,
        },
        ExpiresIn=300 

    )

    s3_url = f"https://{S3_ML_BUCKET_NAME}.s3.amazonaws.com/{key}"

    return presigned_url, s3_url